### PR TITLE
425 support native docker compose api

### PR DESCRIPTION
**Pull request by baez90**
# What does this PR do?

Deprecate shell-escape based `LocalDockerCompose` implementation and provide new `docker/compose` based one using the Docker API. Introduce a new API that takes `context.Context` into account. Furthermore it exposes more typed options.

#425

### CODE REVIEW

**Comment by mdelapenya** [comment]
@baez90 shall we change the PR to draft? We can start the review once you feel confortable with it

And thanks for your time here, using the native Go libraries for compose is a great addition that none of the other languages can do! :kudos:

**Comment by mdelapenya** [comment]
> We can do that if you prefer. I would actually appreciate a first review! To make the pipeline working I'd have to remove Go versions < 1.16. I could do that but wasn't sure if this is okay?

Fine for me about the removal 🙋‍♂️

File: .github/workflows/ci-podman.yml
Comment (baez90) [reply .github/workflows/ci-podman.yml:9]: Unfortunately I haven't had time yet to finish the rebase so GitHub Actions doesn't run at all right now I think 😅

File: compose_api.go
Comment (baez90) [reply compose_api.go]: Makes sense, I updated it accordingly :blush:

File: compose_api.go
Comment (mdelapenya) [reply compose_api.go]: Oh yes, I meant to have the error and the if err != nil in the next line, one next to the other. Personal OCT 🤷

**Comment by mdelapenya** [comment]
Thanks for this PR! Looks super promising! 

It would be also awesome if we improve the docs about using compose with the new API design.

**Comment by baez90** [comment]
Sure, I'm actually wondering if I should try to get a PR into `docker/compose` to update their `docker/docker` dependency because this could resolve some `replace` issues I encountered and which we should otherwise document as well.

I'll give it a shot tomorrow to see if the dependency update would make things easier and update the docs either way accordingly :blush:

**Comment by mdelapenya** [comment]
@baez90 no updates on this PR as I've been very busy with certain updates that will be public soon. I'll let you know once I'm 100% available to continue with this (hopefully soon)

**Comment by baez90** [comment]
Don't you worry I'm a bit busy myself as well. As previously mentioned I tried if upgrading the Docker version would help to get rid of the replacement stuff but doesn't work. Probably when Docker 22.xx is released or so 😅

I'll try to update the docs probably tomorrow or begining of next week and I also have to rebase the branch once more. Do you mind if I add something about Podman support as well when I'm already on it or shall I create another PR for that?

Until then good luck with whatever you're currently working on 😉

**Comment by mdelapenya** [comment]
> Do you mind if I add something about Podman support as well when I'm already on it or shall I create another PR for that?

That would be perfect in a separate PR, faster to review and merge.

> Until then good luck with whatever you're currently working on 😉

Thanks my friend!!

File: .github/workflows/ci-podman.yml
Comment (mdelapenya) [reply .github/workflows/ci-podman.yml:9]: I'm going to apply this change to main branch in a separate PR, as it's impacting other PRs, specially dependabot ones including security fixes. Please see https://github.com/testcontainers/testcontainers-go/issues/492

**Comment by baez90** [comment]
Alright @mdelapenya I prepared at least some docs, rebased everything once more and updated the dependencies.

Let me know if I should update the description once more or if the docs need further adjustments or anything else :blush: till then I'll have a look at #496

**Comment by kiview** [comment]
I just wanted to add a reference to an issue I remembered from the Compose project where someone asked if `compose` can be used as a library and actual API, only to realize that you opened it 🤯 

So for future reference, this is the issue where they track the progress on this:
https://github.com/docker/roadmap/issues/387

I'd suggest only releasing stable support for this, once upstream provides stable support. However, since you already distinguish by API within `tc-go`, it might be ok.

**Comment by baez90** [comment]
Well `tc-go` itself is also not release yet as stable hence I figured it should be okay to use the (fairly) stable `compose` API especially because everything `compose` related is wrapped so breaking changes (at least small-ish ones) could be adapted without breaking the `tc-go` public API.

I'll resolve the conflicts once more and afterwards we can discuss how to proceed with this PR?

**Comment by kiview** [comment]
Maybe `tc-go` is abstracting compose further away as `tc-java` does it, but Compose V2 has some pretty significant breaking changes, as we found while working on https://github.com/testcontainers/testcontainers-java/pull/5608 (@eddumelendez can probably share the concrete details here). 

Most prominent is the container/service naming, using hyphens instead of underscores.

Regarding:
> Well tc-go itself is also not release yet as stable

while technically true, every breaking change will reflect on how users (and there are existing users of `tc-go` of course) perceive the maturity of the library, completely independent on what is actually communicated in the version number.

**Comment by baez90** [comment]
Funny that you mention the hyphens-vs-underscores problem as I took care of that a while ago :smile: (see [here](https://github.com/testcontainers/testcontainers-go/blob/main/compose.go#L35) if you're interested) when I noticed that most of the test code was still running with compose V1 while I was running the tests locally with V2 but apart from that I haven't experienced any further problems with compose V1 vs. V2.

I totally understand your concern that the API should be as stable as possible to ensure `tc-go` users don't get frustrated by having to deal with breaking changes every time they update. I am using it on a daily basis myself and my first contribution was actually to rework a previous 'non-breaking-change' that kept the existing API but changed the behavior regarding bind mounts with the result that the code was still compiling but my tests were failing because mount source and target were inverted (previously the map key was the source and the value the target and suddenly it was the other way around) which is a great example that sometimes breaking the API is better than keeping it :smile: 

But as I said previously the new public API does not directly expose any compose details and also only includes some basic compose options so I'm pretty confident almost every breaking change can be hidden. The worst case I can think of is that some options aren't supported in the future and this can still be handled gracefully by deprecating the `tc-go` option and make it a noop option for a few releases.

Do I miss something?

**Comment by eddumelendez** [comment]
just want to add that besides the hyphens/underscore change in compose v2 the compose file version supported in v2 starts with 2. There should be `services` at the root of the yaml file

**Comment by baez90** [comment]
Is something missing here @mdelapenya ?

**Comment by mdelapenya** [comment]
> Is something missing here @mdelapenya ?

Hey @baez90 we are preparing the next release (0.14.0) and would like to have this refactor of docker-compose ready for the next release (0.15.0), so until we set up the basis for a more formal release process, I'll postpone merging this until the 0.14.0 release is out (really soon)

Therefore we will be able to deliver these great changes, using the native Go APIs for compose, in a separate release without impacting existing users.

This is my plans, that I wanted to share with you before doing anything else.

File: compose_api.go
Comment (marcpfuller) [inline compose_api.go:208]: I think this should return the interface "Container" instead of DockerContainer

Comment (baez90) [reply compose_api.go:208]: Go's best practice is normally 'accept interfaces, return structs' therefore I think it should be fine to return the `*DockerContainer` and not `Container`

File: compose_local.go
Comment (mdelapenya) [inline compose_local.go:121]: @baez90 we are going to merge https://github.com/testcontainers/testcontainers-go/pull/514, and I think this line would need to be updated.

Can you double check this line: https://github.com/testcontainers/testcontainers-go/pull/514/files#diff-4c4f2f19edfc5e047aef7e584ae2608724405f66563c6ee3b9957a19febada2bR183

Once merged, could you double check if the new test in it should be updated? 🙏

Comment (baez90) [reply compose_local.go:121]: Do you mind if I rebase my PR? I'll resolve the conflicts and make sure the tests are working again and force push afterwards without new changes?

File: docker.go
Comment (mdelapenya) [inline docker.go:646]: Should we update the usage of client.Client elsewhere? https://github.com/testcontainers/testcontainers-go/pull/476/files#diff-b68159b6168f099a639e8c6f96d75f23333ab0f0505ee22a98268748db17ae4cR705

File: compose_api.go
Comment (mdelapenya) [inline compose_api.go:111]: It would be great if we add comments for the attributes, so it's usage is even simpler

File: compose_api.go
Comment (mdelapenya) [inline compose_api.go:247]: could we explain in a comment the need of increasing the capacity by two elements?

**Review by mdelapenya — COMMENTED** [review]
I added a few comments regarding consistency (client.APIClient Vs client.Client), naming (NewDockerCompose Vs NewDockerComposeAPI) and adding comments for clarity.

I added a suggestion on how to configure the docker client (client.APIClient) reading the testcontainers.properties file, as we need this to better support remote Docker environments.

Other than that, it works fine in my local machine (Mac M1), will wait for your feedback to provide a final ✅ 

Great job!

File: compose.go
Comment (baez90) [reply compose.go:32]: I previously did that but having a more verbose name makes sure you can implement multiple option interfaces on a single type like:

``go
type RemoveOrphans bool

func (ro RemoveOrphans) applyToStackUp(o *stackUpOptions) {
	o.RemoveOrphans = bool(ro)
}

func (ro RemoveOrphans) applyToStackDown(o *stackDownOptions) {
	o.RemoveOrphans = bool(ro)
}
```

this way it's a bit more flexible :wink:

Comment (mdelapenya) [reply compose.go:32]: I see. Let's keep it as is. Thanks for the clarification

File: compose.go
Comment (baez90) [reply compose.go:84]: Sure, I just didn't want to assume this would be the default :see_no_evil:

File: compose_api.go
Comment (baez90) [reply compose_api.go:171]: you're absolutely right, will split it. The original version's a bit much at once :smile:

File: compose.go
Comment (baez90) [reply compose.go:100]: tbh, I already thought about it and figured I could add this later :smile: but this is a lot better, thanks!

Comment (mdelapenya) [reply compose.go:100]: Yeah, it will reuse all the optimizations that we can support from the testcontainers.properties file

File: compose_api.go
Comment (baez90) [reply compose_api.go:247]: I'll introduce a constant with a speaking name, hope that does the trick for you as well :wink:

**Comment by baez90** [comment]
I went through all the comments and applied them (almost) exactly as suggested, let me know if you find anything else you want me to change :blush:

File: compose_api.go
Comment (mdelapenya) [inline compose_api.go:132]: Loved this line! ❤️

**Review by mdelapenya — COMMENTED** [review]
LGTM although the CI is complaining. Could you take a look? Once resolved, I'm giving the 👍

**Comment by baez90** [comment]
I'll have a look next week, just noticed, that the CI isn't green anymore :see_no_evil:

**Comment by mdelapenya** [comment]
I believe the CI errors come from the locks and the wait strategies 🤔 

```shell
goroutine 39 [semacquire, 9 minutes]:
sync.runtime_Semacquire(0xc0002944e0?)
	/opt/hostedtoolcache/go/1.18.6/x64/src/runtime/sema.go:56 +0x25
sync.(*WaitGroup).Wait(0xc0004fc240?)
	/opt/hostedtoolcache/go/1.18.6/x64/src/sync/waitgroup.go:136 +0x52
golang.org/x/sync/errgroup.(*Group).Wait(0xc0002cc4c0)
	/home/runner/go/pkg/mod/golang.org/x/sync@v0.0.0-20220819030929-7fc1605a5dde/errgroup/errgroup.go:53 +0x27
github.com/testcontainers/testcontainers-go.(*dockerCompose).Up(0xc0004f6630, {0x16c2d00, 0xc000090d00}, {0xc0003f26a0, 0x1, 0x171f876?})
	/home/runner/work/testcontainers-go/testcontainers-go/compose_api.go:226 +0x7c5
github.com/testcontainers/testcontainers-go.TestDockerComposeAPIStrategyForInvalidService(0xc000294000)
	/home/runner/work/testcontainers-go/testcontainers-go/compose_api_test.go:43 +0x212
testing.tRunner(0xc000294000, 0x1535a90)
	/opt/hostedtoolcache/go/1.18.6/x64/src/testing/testing.go:1439 +0x102
created by testing.(*T).Run
	/opt/hostedtoolcache/go/1.18.6/x64/src/testing/testing.go:1486 +0x35f
```

**Comment by baez90** [comment]
You were right, I missed one call to a public function I added a lock call to, to avoid concurrent read-write ops to the container cache. I made this one private, removed the locks and added a public wrapper so that the locks are still in place but the second occurrence is using the private implementation :blush:

**Comment by baez90** [comment]
Rebased the branch once more onto `main` and did a final dependency update for `docker/compose`. As long as there's nothing more, I'd say this is ready :smile:

**Comment by mdelapenya** [comment]
Weird: seeing this error in mac when running `go vet` https://github.com/testcontainers/testcontainers-go/actions/runs/3191143864/jobs/5207098574

**Comment by baez90** [comment]
yeah, I upgraded to latest `docker/compose` and now both `CreateOptions` and `StartOptions` have a `Services` field :roll_eyes:

**Comment by baez90** [comment]
Hopefully this works now, otherwise I'll fix it tomorrow evening

**Comment by mdelapenya** [comment]
I think this is ready. Let me do a last minute review, but other than that, LGTM. Huge PR, profound changes, big thanks

**Comment by mdelapenya** [comment]
> 😄 it's not really weird, the merge removed the `replace` directives in the `go.mod` currently necessary to work around the problem you noticed until `moby/moby` will probably finally release a version with Go modules support and a newer version of `github.com/tonistiigi/fsutil` 🙄

Ah thanks! The merge conflicts resolution on mod files is many times a mess 🤦  We merged #548, which included a run of `go mod tidy`, and that caused it. I'm tempted to add a pre-commit hook to validate the go mod, so it's not run on CI only

File: docs/features/docker_compose.md
Comment (mdelapenya) [inline docs/features/docker_compose.md:22]: Double checking the docs I found that we did not update them when we decided to rename the function. Could you verify the example code is executable? 🙏 
```suggestion
Use the convenience `NewDockerCompose(...)` constructor which creates a random identifier and takes a variable number
```

File: docs/features/docker_compose.md
Comment (mdelapenya) [inline docs/features/docker_compose.md:165]: We lost the indentation here

File: docs/features/docker_compose.md
Comment (mdelapenya) [inline docs/features/docker_compose.md:168]: Same indentation issue

File: docs/features/docker_compose.md
Comment (mdelapenya) [inline docs/features/docker_compose.md:186]: Same indentation issue

File: docs/features/docker_compose.md
Comment (mdelapenya) [inline docs/features/docker_compose.md:147]:
```suggestion
```

File: docs/features/docker_compose.md
Comment (mdelapenya) [inline docs/features/docker_compose.md:12]: Ups, we almost missed this
```suggestion
```

**Review by mdelapenya — APPROVED** [review]
LGTM, I'm approving the PR in advance although I left last minute comments in the docs regarding the name of the `NewDockerComposeAPI` function, and a not resolved merge conflict. 

It took a while to review, but the changes are so profound that the time was needed. I will take a note to create an issue to remove the local compose code in upcoming releases.

Super good PR including lots of tests for the Compose native API. Again, thanks for your commitment to testcontainers-go! 🚀

**Comment by baez90** [comment]
I incorporated the docs changes we spoke of and added another section regarding the service name API.
Also updated to the latest `docker-compose` version and rebased the branched once more to resolve conflicts.
Let me know if there's anything else missing :wink:

**Comment by mdelapenya** [comment]
At the moment there is a force-push, I lose track of what I'd already reviewed and what I hadn't. I'd thank if we avoid pushing-force the PRs 🙏

File: docs/features/docker_compose.md
Comment (mdelapenya) [inline docs/features/docker_compose.md:34]:
```suggestion
	assert.NoError(t, err, "NewDockerCompose()")
```

File: docs/features/docker_compose.md
Comment (mdelapenya) [inline docs/features/docker_compose.md:65]:
```suggestion
	assert.NoError(t, err, "NewDockerComposeWith()")
```

File: docs/features/docker_compose.md
Comment (mdelapenya) [inline docs/features/docker_compose.md:82]:
```suggestion
To interact with service containers after a stack was started it is possible to get a `*tc.DockerContainer` instance via the `ServiceContainer(...)` function.
```

File: docs/features/docker_compose.md
Comment (mdelapenya) [inline docs/features/docker_compose.md:86]:
```suggestion
Furthermore, there's the convenience function `Services()` to get a list of all services **defined** by the current project.
```

**Comment by baez90** [comment]
I added another commit regarding the docs yesterday and did not change any previous commit, I only rebased it onto `main` again to resolve conflicts :wink: no rewrite of the history happened, you only need to review the latest commit :blush: hope that helps

File: docs/features/docker_compose.md
Comment (mdelapenya) [inline docs/features/docker_compose.md:148]:
```suggestion
_Node:_ this API is **deprecated** and superseded by `ComposeStack` which takes advantage of `docker-compose` v2 being
```

**Review by mdelapenya — COMMENTED** [review]
Minor comments or typos in docs

**Comment by mdelapenya** [comment]
> If you prefer I'll merge `main` if necessary in the future to resolve conflicts, I'm just used to the rebase workflow 😉

Yes please 🙏  that would help!

Apart from that, the CI builds are failing, seems related to `TestDockerCreateContainerWithDirs`. Could you take a look?

**Comment by mdelapenya** [comment]
@baez90 I found https://github.com/testcontainers/testcontainers-go/pull/574, which it's fixing the CI

**Comment by mdelapenya** [comment]
@baez90 I've noticed this error in the pipeline, related to a concurrent access to a map (compose_api.go:294):

```
=== FAIL: . TestDockerComposeApiWithWaitForShortLifespanService (unknown)
Network b2e2bce1-ac00-48f3-82fc-8403274bef5c_default  Creating
Network b2e2bce1-ac00-48f3-82fc-8403274bef5c_default  Created
Container b2e2bce1-ac00-48f3-82fc-8403274bef5c-falafel-1  Creating
Container b2e2bce1-ac00-48f3-82fc-8403274bef5c-tzatziki-1  Creating
Container b2e2bce1-ac00-48f3-82fc-8403274bef5c-tzatziki-1  Created
Container b2e2bce1-ac00-48f3-82fc-8403274bef5c-falafel-1  Created
Container b2e2bce1-ac00-48f3-82fc-8403274bef5c-falafel-1  Starting
Container b2e2bce1-ac00-48f3-82fc-8403274bef5c-tzatziki-1  Starting
Container b2e2bce1-ac00-48f3-82fc-8403274bef5c-falafel-1  Started
Container b2e2bce1-ac00-48f3-82fc-8403274bef5c-tzatziki-1  Started
fatal error: concurrent map writes

goroutine 710 [running]:
runtime.throw({0x1f5e0d1?, 0x70?})
	/opt/hostedtoolcache/go/1.18.7/x64/src/runtime/panic.go:992 +0x71 fp=0xc0004e7c78 sp=0xc0004e7c48 pc=0x43a631
runtime.mapassign_faststr(0x100?, 0x0?, {0x1f45191, 0x7})
	/opt/hostedtoolcache/go/1.18.7/x64/src/runtime/map_faststr.go:212 +0x39c fp=0xc0004e7ce0 sp=0xc0004e7c78 pc=0x41551c
github.com/testcontainers/testcontainers-go.(*dockerCompose).lookupContainer(0xc000865200, {0x2277d08, 0xc00041d140}, {0x1f45191, 0x7})
	/home/runner/work/testcontainers-go/testcontainers-go/compose_api.go:294 +0x4fe fp=0xc0004e7f20 sp=0xc0004e7ce0 pc=0x19f65de
github.com/testcontainers/testcontainers-go.(*dockerCompose).Up.func1()
	/home/runner/work/testcontainers-go/testcontainers-go/compose_api.go:229 +0x57 fp=0xc0004e7f78 sp=0xc0004e7f20 pc=0x19f5a77
golang.org/x/sync/errgroup.(*Group).Go.func1()
	/home/runner/go/pkg/mod/golang.org/x/sync@v0.1.0/errgroup/errgroup.go:75 +0x64 fp=0xc0004e7fe0 sp=0xc0004e7f78 pc=0x62da84
runtime.goexit()
	/opt/hostedtoolcache/go/1.18.7/x64/src/runtime/asm_amd64.s:1571 +0x1 fp=0xc0004e7fe8 sp=0xc0004e7fe0 pc=0x46da61
created by golang.org/x/sync/errgroup.(*Group).Go
	/home/runner/go/pkg/mod/golang.org/x/sync@v0.1.0/errgroup/errgroup.go:72 +0xa5
```

Could you take a look? 🙏

**Comment by mdelapenya** [comment]
@baez90 test are passing! I pushed https://github.com/testcontainers/testcontainers-go/pull/476/commits/4111cdcba9abc102b9079b02f8256ab4479c729f and https://github.com/testcontainers/testcontainers-go/pull/476/commits/b231f60d3bbe20d1464c64e6a1f8884ac213a8c5 with the sole goal of making this PR into main ASAP. Hope you understand.

My plans are: release v0.15.0 this week(end) without this PR, and immediately merge this PR for a dedicated v0.16.0 to isolate the compose changes. Does it sound good to you?

**Comment by mdelapenya** [comment]
@baez90 I'm merging this PR now. Big big thanks for the huge refactor which simplifies a lot the maintenance of the compose module. We will eventually remove (and communicate) the local compose code after having the native implementation.

**Comment by baez90** [comment]
Thanks for taking care! I just couldn't manage to get some time to work on this PR so thanks a lot! Also for all the support, time for reviews and your valuable input all the time!

**Pull request merged**
